esphome/components/samsung/samsung.cpp

```diff
@@ -0,0 +1,238 @@
+#include "samsung.h"
+
+namespace esphome {
+namespace samsung {
+
+void SamsungClimate::transmit_state() {
+  if (current_climate_mode_ != climate::ClimateMode::CLIMATE_MODE_OFF &&
+      this->mode == climate::ClimateMode::CLIMATE_MODE_OFF) {
+    set_and_send_power_state_(false);
+    return;
+  }
+
+  if (current_climate_mode_ == climate::ClimateMode::CLIMATE_MODE_OFF &&
+      this->mode != climate::ClimateMode::CLIMATE_MODE_OFF) {
+    set_and_send_power_state_(true);
+  }
+
+  current_climate_mode_ = this->mode;
+
+  set_mode_(this->mode);
+  set_temp_(this->target_temperature);
+  set_swing_(this->swing_mode);
+  set_fan_(this->fan_mode.has_value() ? this->fan_mode.value() : climate::CLIMATE_FAN_AUTO);
+
+  send_();
+}
+
+/// Send the current state of the climate object.
+void SamsungClimate::send_() {
+  checksum_();
+
+  auto transmit = this->transmitter_->transmit();
+  auto *data = transmit.get_data();
+  data->set_carrier_frequency(SAMSUNG_IR_FREQUENCY);
```

Comment:
```suggestion
  // Header(2) +  2 * MSG_HDR(2 * Item(2)) + 21 Bytes * 8 Bits * Bit(2) + Last Mark (1)
  data->reserve(2 + 2 * 2 * 2 + 21 * 8 * 2 + 1);
  data->set_carrier_frequency(SAMSUNG_IR_FREQUENCY_HZ);
```

That is 347 items for dynamic size array. Mark or space take up 1 element of space in vector. Item is 2 elements. I think that this is correct but take a look at code and check it. Should be faster at runtime if its reserved.